### DISCUSSION TITLE
[WEC-Sim Applications] OWC (7DOF) Application in WEC-Sim BEMIO/CAPYTAINE

### DISCUSSION THREAD

**Discussion by dforbush2** [WEC-Sim/WEC-Sim#1294] [Q&A] [answered]
Opened on behalf of @samir-miloud

**Cite the [WEC-Sim applications](https://github.com/WEC-Sim/WEC-Sim_Applications) case in question**
OWC applications case

**Describe the bug or issue**
The OWC applications case has 7 DOF, the 7th degree of freedom corresponding to the heave motion of the internal free surface. How is this achieved using Capytaine?

**To Reproduce**
Look at the *h5 file for the OWC model.

**Expected behavior**
A Capytaine run using 7 degrees of freedom.

**Comment by dforbush2** [comment]
Hi @samir-miloud

As you've noticed, the OWC applications case is a 7DOF model of an oscillating water column, using a modified flexible body block. Now your question seems to be how to develop the 7 DOF set of BEM coefficients using Capytaine,

Please look at 
1) Assigning a custom degree of freedom: https://capytaine.github.io/stable/user_manual/cookbook.html#custom-degree-of-freedom

2). Definition of generalized body modes for RM3: https://github.com/WEC-Sim/WEC-Sim/blob/main/examples/BEMIO/CAPYTAINE/rm3/gbm_dofs.py

3). Applying generalized body modes to a floating body: https://github.com/WEC-Sim/WEC-Sim/blob/main/examples/BEMIO/CAPYTAINE/call_capytaine.py#L212
Thanks @akeeste for help on this issue.

**Comment by samir-miloud** [comment]
Hi @dforbush2,

I've followed your advice and successfully added the 7th degree of freedom (DoF) for the oscillating water column (OWC). I assumed that the OWC has 6 DoFs (Heave, Roll, Pitch, etc.) plus an additional DoF for the column height. Using Capytaine for the hydrostatic calculations, I generated the file "OWC_spar.h5" using bemio.m, which contains the necessary matrices (Mass Matrix, Stiffness Matrix, Damping Matrix, Added Mass Matrix, Diffraction Matrix, Reaction Matrix).
The results from running bemio.m are included in the attached "Results.zip" file. This file contains the output for each matrix and other relevant data.
However, when I try to run the simulation in WEC-Sim, I encounter the following error message:

">> addWecSimSource
>> wecSim
WEC-Sim: An open-source code for simulating wave energy converters
Version: 6.0
Initializing the Simulation Class...
	Case Dir: /home/samir/Documents/GitHub/WEC-SIM/WEC-Sim_Applications/OWC0/OrificeModel 

WEC-Sim Input From Standard wecSimInputFile.m Of Case Directory... 
WEC-Sim: An open-source code for simulating wave energy converters
Version: 6.0

Initializing the Simulation Class...
	Case Dir: /home/samir/Documents/GitHub/WEC-SIM/WEC-Sim_Applications/OWC0/OrificeModel  
Elapsed time is 4.013673 seconds.
WEC-Sim Pre-processing ...   
**Unrecognized field name "gbm".**

Error in bodyClass/hydroForcePre (line 356)
                obj.hydroForce.gbm.stiffness=obj.hydroData.gbm.stiffness;

Error in initializeWecSim (line 287)
        body(it).hydroForcePre(waves(1).omega,waves(1).direction,simu.cicTime,waves(1).bem.count,simu.dt,...

Error in run (line 91)
evalin('caller', strcat(script, ';'));

Error in wecSim (line 43)
run('initializeWecSim');"


[Results.zip](https://github.com/user-attachments/files/16630037/Results.zip)

I have attached both the "OWC_spar.h5" file and the "Results.zip" file for your reference.

Could you please help me resolve this issue?

Thank's

Best regards,
Samir

**Comment by dforbush2** [comment]
Hi Samir,
It seems like your h5 file is missing, and that is the one I really need to see to understand the error. Looking at your *zip files, things look OK except for the radiation damping in heave and GBM, which drops negative for a few frequencies. I'd check out badBEMFixfcn on the development branch for some help with that.

Best,
Dom

**Comment by samir-miloud** [comment]
Hi Dom,

Thank you very much for your prompt feedback and advice. I will check the
radiation damping in heave and GBM as you suggested and explore
badBEMFixfcn on the development branch.
I'm also attaching the h5 file as requested.

Thanks again for your valuable help.

Best regards,
Samir

**Comment by samir-miloud** [comment]
Hi @dforbush2,

I am still encountering the same error we previously discussed in my WEC-Sim simulation. Since our last conversation, I have explored the bemio.m and writeBEMIOH5.m scripts to try to understand the origin of the problem.
It seems that the error is related to the "Generalized Body Modes" (GBM). Specifically, I suspect that the issue lies in the generation or inclusion of the GBM coefficients in the final file OWC_spar.h5. I have noticed that these coefficients should be present in the output file, but in my case, they are not.
Now, I am unsure where these coefficients should be calculated: is it in the .nc file for Capytaine, in the .out file for WAMIT, or by another function in bemio.m?

Thanks!
[OWC_spar.zip](https://github.com/user-attachments/files/16671129/OWC_spar.zip)

**Comment by dforbush2** [comment]
Hi @samir-miloud 

Your h5 file looks nice! Can you please share your full run directory with your input file, your bemio.m call, and the *.nc files from Captyaine? I am wondering if there is a bug in the capytaine processing scripts of bemio that are not properly recognizing that the additional modes are GBM modes, but this will be much easier for me to do if you can share the requested files. If there are concerns about any of the files being public please let me know and we can work something out.

Best,
Dom

**Comment by samir-miloud** [comment]
Hi @dforbush2,
Thank you for your feedback. I’ve attached the full run directory, including the input file, the bemio.m call, and the *.nc files generated by Capytaine, as requested.
If you need any additional information or if there are any issues with the files, please let me know. I’m open to discussing any alternative solutions if certain files need to remain confidential.
Thanks.
Best regards,
[Uploading OWC.zip…]()

Samir

**Comment by samir-miloud** [comment]
Hi @dforbush2,

I hope you’re doing well. I wanted to follow up on the files I sent a few days ago and share an observation I’ve made since.

While working on the GBM issue, I noticed a key difference between the 'readCAPYTAINE.m' and 'readWAMIT.m' functions in 'bemio.m'. Specifically, the 'readWAMIT.m' function calculates the 'hydro(F).gbm' coefficients, whereas 'readCAPYTAINE.m' does not. This seems to be the reason why my '.h5' file lacks the GBM data when using Capytaine, while WAMIT successfully includes it.

It seems that using WAMIT might be the most straightforward way to obtain the complete '.h5' file for the OWC case. However, if the issue indeed lies with 'readCAPYTAINE.m', I would greatly appreciate your help in improving this function to include GBM, similar to 'readWAMIT.m'.

Thanks

Best regards,
Samir

**Comment by dforbush2** [comment, marked as answer]
Hi @samir-miloud 

my apologies for the delay, i've been ill and unable to monitor this discussion. Also, please note that your last upload of the *zip file appears to have failed (or perhaps you posted this issue prior to the upload completing?)  Please re-try this: as noted below this will help me out a lot. 

I think you are correct with the readCapytaine file...I will open an issue on your behalf on our issues page highlighting this issue regarding the missing gbm field, and, as noted below, the improperly sized hydroData matrix. 

However, do note that the hydro.gbm coefficients are used to define the inertia, stiffness, and damping of the generalized body mode. In your  case, the appropriate values for all of these are zero. This is a bit confusing, but these parameters are intended to be any inertia, stiffness, and dampings NOT the result of hydrodynamics (already captured in the 7th degree of freedom of radiation damping, added mass, excitation, hydrostatic stiffness, etc). For example, consider an OWC that has a rubber sheet over the orifice that resists deformation with some non-zero mass, stiffness, and damping: these are the values that should go in the gbm fields, since these forces act in addition to the mass, stiffness, and damping contributed from hydrostatics. 

The one other issue I see is that the hydrostatic stiffness matrix ([data].hydro_coeffs.linear_restoring_stiffness) is 6 x 6 in your case where you would prefer it to be 7 x 7.  Additionally and oddly, these values are all 0 for you, when they should be non-zero for (at a minimum) element (3,3), (4,4), (5,5), and your missing (7,7). I also suspect you would expect nonzero stiffnesses in (3,7) and (7,3). I suspect there is an issue reading the capytaine files (another problem in readCapytaine.nc) or in your BEM setup...so please do re-try that *zip file upload so I can tell which one!

Capytaine may not output a stiffness matrix larger than 6 x 6...WAMIT doesn't do this either. If this is the case, I would manually edit the generated h5 file to add a 7th row and column to capture the hydrostatic stiffness of your internal water column. Linearized, this is equivalent for (3,7), (7,7), and (7,3) and is equal to the free surface area of the water column (remember the NORMALIZED values are in the h5 file).  

Best,
Dom

**Comment by samir-miloud** [comment]
Hi Dom,

I would like to sincerely thank you for your detailed feedback and
invaluable guidance. Following your instructions, I was able to resolve the
issues and get the code running successfully.

I have attached a zip file containing the results of my simulation,
including the following files:

- OWC.png (visualization of the results),
- OWC.avi (simulation animation),
- OWC_spar.h5 (generated h5 file).
Once again, I greatly appreciate your help and support. If it's alright
with you, could I reach out to you in the future if I encounter further
difficulties?

Thanks,
Samir
